arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,2917 @@
+       * INT_STAT reg.
+       */
+
+      if (!isr)
+        {
+          return irq_loops ? OK : -1;
+        }
+
+      /* Receive Buffer Not Empty Interrupt */
+
+      if (isr & MPFS_CANFD_INT_STAT_RBNEI)
+        {
+          /* Mask RBNEI first, then clear interrupt. Even if
+          * another IRQ fires, RBNEI will always be 0 (masked).
+          */
+
+          icr = MPFS_CANFD_INT_STAT_RBNEI;
+          putreg32(icr, priv->base + MPFS_CANFD_INT_MASK_SET_OFFSET);
+          putreg32(icr, priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+
+          /* Schedule work to process RX frame from CAN controller */
+
+          work_queue(CANWORK, &priv->rxwork, mpfs_receive_work, priv, 0);
+        }
+
+      /* TXT Buffer HW Command Interrupt */
+
+      if (isr & MPFS_CANFD_INT_STAT_TXBHCI)
+        {
+          /* Clear TX interrupt flags */
+
+          mpfs_txdone(priv);
+
+          /* Schedule work to poll for next available tx frame from the
+           * network.
+           */
+
+          work_queue(CANWORK, &priv->txdwork, mpfs_txdone_work, priv, 0);
+        }
+
+      /* Error Interrupts */
+
+      if (isr & MPFS_CANFD_INT_STAT_EWLI ||
+          isr & MPFS_CANFD_INT_STAT_FCSI ||
+          isr & MPFS_CANFD_INT_STAT_ALI ||
+          isr & MPFS_CANFD_INT_STAT_BEI)
+        {
+          icr = isr & (MPFS_CANFD_INT_STAT_EWLI |
+                      MPFS_CANFD_INT_STAT_FCSI |
+                      MPFS_CANFD_INT_STAT_ALI |
+                      MPFS_CANFD_INT_STAT_BEI);
+          canerr("Some error interrupts. Clearing 0x%08x\n", icr);
+          putreg32(icr, priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+          mpfs_err_interrupt(priv, isr);
+        }
+    }
+
+  /* Now, it seems that there are still some interrupt flags that remain
+   * stuck in INT_STAT reg.
+   */
+
+  canerr("Stuck interrupt (isr=%08x)\n", isr);
+
+  /* Check if any of the stuck one belongs to txb status. */
+
+  if (isr & MPFS_CANFD_INT_STAT_TXBHCI)
+    {
+#if 0
+      candebug("txb_sent=0x%08x txb_processed=0x%08x\n", priv->txb_sent,
+               priv->txb_processed);
+      for (int i = 0; i < priv->ntxbufs; i++)
+        {
+          uint32_t status = mpfs_can_get_txb_status(priv, i);
+          candebug("txb[%d] txb status=0x%08x\n", i, status);
+        }
+#endif
+      /* Notify to socket interface */
+
+      NETDEV_TXERRORS(&priv->dev);
+
+      /* Clear txb status change interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_TXBHCI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Check if any of the stuck one belongs to RX buffer data overrun */
+
+  if (isr & MPFS_CANFD_INT_STAT_DOI)
+    {
+      /* Notify to socket interface */
+
+      NETDEV_RXERRORS(&priv->dev);
+
+      /* Clear Data Overrun interrupt */
+
+      putreg32(MPFS_CANFD_INT_STAT_DOI,
+               priv->base + MPFS_CANFD_INT_STAT_OFFSET);
+    }
+
+  /* Clear and reset all interrupt. */
```

Comment:
> I don't quite understand why the interrupt routine has to manually disable the interrupt source, will this not result in missed interrupts ? Where does the source get re-enabled ?

-> this only happens in rare error scenario when there is stuck interrupt which still remains after the upper limit of the irq loop number (10000) is exceeded. In such case, we reset all interrupt and start afesh.